File: C++/Sorting Algorithms/quick_Sort.cpp
// Program to perform the quick sort
#include<iostream>

using namespace std;

// function to scan the array
void scan_Array(int array_1[], int size)
{
    cout << "Enter the elements of the array: \n";
    for(int i=0; i<size; i++)
    {
        cout << "Element [" << i << "]: ";
        cin >> array_1[i];
    }
}

// function to print the array
void print_Array(int array_1[], int size)
{
    for(int i=0; i<size; i++)
    {
        cout << array_1[i] << " ";
    }
}

// function to swap the elements
void swap_Elements(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

// function to perform partition in array
int partition(int array_1[], int p, int r)
{
    int i = p - 1;
    int x = array_1[r];
    for(int j = p; j < r; j++)
    {
        if(array_1[j] <= x)
        {
            i++;
            swap_Elements(&array_1[i], &array_1[j]);
        }
    }

    swap_Elements(&array_1[i+1], &array_1[r]);

    return i+1;
}

// function to perform quick sort
void quicksort(int array_1[], int p, int r)
{
    if(p < r)
    {
        int q = partition(array_1, p, r);
        quicksort(array_1, p, q-1);
        quicksort(array_1, q+1, r);
    }
}

int main()
{
    int size, array_1[100];

    // scanning the size of the array
    cout << "Enter the size of the array: ";
    cin >> size;

    // scanning the elements of the array
    scan_Array(array_1, size);

    // assigning the zeroth index to p
    int p = 0;

    // assigning the last index to r
    int r = size-1;

    // printing the scanned array
    cout << "\nArray: \n";
    print_Array(array_1, size);

    // calling the quicksort function
    quicksort(array_1, p, r);

    // printing the sorted array
    cout << "\nSorted Array: \n";
    print_Array(array_1, size);

    return 0;
}

File: JavaScript/Sorting Algorithms/quick_sort.js

// Javascript implementation of QuickSort


// A utility function to swap two elements
function swap(arr, i, j) {
	let temp = arr[i];
	arr[i] = arr[j];
	arr[j] = temp;
}

/* This function takes last element as pivot, places
the pivot element at its correct position in sorted
array, and places all smaller (smaller than pivot)
to left of pivot and all greater elements to right
of pivot */
function partition(arr, low, high) {

	// pivot
	let pivot = arr[high];

	// Index of smaller element and
	// indicates the right position
	// of pivot found so far
	let i = (low - 1);

	for (let j = low; j <= high - 1; j++) {

		// If current element is smaller
		// than the pivot
		if (arr[j] < pivot) {

			// Increment index of
			// smaller element
			i++;
			swap(arr, i, j);
		}
	}
	swap(arr, i + 1, high);
	return (i + 1);
}

/* The main function that implements QuickSort
		arr[] --> Array to be sorted,
		low --> Starting index,
		high --> Ending index
*/
function quickSort(arr, low, high) {
	if (low < high) {

		// pi is partitioning index, arr[p]
		// is now at right place
		let pi = partition(arr, low, high);

		// Separately sort elements before
		// partition and after partition
		quickSort(arr, low, pi - 1);
		quickSort(arr, pi + 1, high);
	}
}

// Function to print an array
function printArray(arr, size) {
	for (let i = 0; i < size; i++)
		document.write(arr[i] + " ");

	document.write("<br>");
}

// Driver Code

let arr = [10, 7, 8, 9, 1, 5];
let n = arr.length;

quickSort(arr, 0, n - 1);
document.write("Sorted array: <br>");
printArray(arr, n);

// This code is contributed by Saurabh Jaiswal

